lib/index.js

```diff
@@ -142,11 +142,7 @@ export class VFile {
     /* eslint-enable no-unused-expressions */
 
     // Set path related properties in the correct order.
-    let index = -1
-
-    while (++index < order.length) {
-      const prop = order[index]
-
+    for (const prop of order) {
```

Comment:
squashes

```
#30. [Survived] EqualityOperator
lib/index.js:147:12
-       while (++index < order.length) {
+       while (++index <= order.length) {
Ran all tests for this mutant.
```

Reply:
I don’t get what this solves 🤷‍♂️ Why would a broken loop suddenly be created.

Reply:
Mutation testing is testing the tests.
Here it found that the condition in the loop could be modified without breaking a test, a signal that there is potential code.
We could put an ignore on the line https://github.com/stryker-mutator/stryker-js/blob/bfb33030b10f8c95fd5a7bb1a9307881b50629f4/docs/disable-mutants.md#using-a--stryker-disable-comment
But here it felt cleaner to use a `for of` loop instead, since the intent is to loop over the elements in the list.

Reply:
To expand, on this change: it’s the essence of why I’m 🤷‍♂️ / :-1: on this PR / mutation tests.

- What Stryker reports is confusing: for me personally, but presumably also for anyone that wants to contribute
- This “solution” is *different*, it doesn’t solve anything except for Stryker

I’m fine with all the added test cases, somewhat verbose, but why not.
But this change is IMO a regression that’s `stryker ignore`ing problems in Stryker itself.

For Stryker more generally: the added tooling to me feels like the direction the MDX repo was heading into a while ago. Too much tooling, too slow, constantly broken and unusable, and hard to actually contribute code to.

I feel like TypeScript and the xo direction already result in much of that. A lot of busy work / churn without benefits.

---

> This is basically a masked `for` loop.

If `xo` starts disallowed `while` loops I would personally yeet it into the trash on every project I can.
I find `while` loops are the most powerful and clear loops that exist. The essence of a loop. `for of` loops have their use but IMO don’t need to be used solely.

Reply:
> What Stryker reports is confusing: for me personally, but presumably also for anyone that wants to contribute

Is there anything that could be added to make it clearer?

In other conversations, we've touched on the topic that mutation testing, can be thought of roughly as a type of test coverage.
If a contributor came through and said "line coverage is confusing, so I didn't add it" you'd close the PR without further discussion.

> This “solution” is different, it doesn’t solve anything except for Stryker

It improves readability, Stryker drew my attention to this line. I would have suggested this change regardless of the coverage difference.

> But this change is IMO a regression

I disagree, readable code is not a regression.

> that’s stryker ignore ing problems in Stryker itself.

Not really, the mutation test is doing what it's supposed to, seeing if editing code breaks things.

> Too much tooling, too slow, constantly broken and unusable,

breaking this down, and focusing this back on vfile, rather than pent up frustration with mdx

* too much tooling: this is highly subjective and I respectfully disagree
* too slow: it takes a minute and thirty seconds total for all tests including mutation tests
* constantly broken: this is a type of test coverage, if the code isn't covered by tests, the build should be broken :thinking: 
* unusable: again highly subjective, I find both vfile and mdx are quite usable

> and hard to actually contribute code to.

in MDX that was more governance related than tooling related

Reply:
Are Stryker results predictable? As in, does it always make the same mutations? 

In my experience it’s frustrating if CI fails due to factors other than the changes made.

Reply:
> Are Stryker results predictable? As in, does it always make the same mutations?

It does have predictable results, it has a fixed number of mutators https://stryker-mutator.io/docs/mutation-testing-elements/supported-mutators
and it runs all enabled mutators through all replacement combinations, on every enabled file.

Reply:
> Is there anything that could be added to make it clearer?

istanbul/c8/etc reports are fine.

> In other conversations, we've touched on the topic that mutation testing, can be thought of roughly as a type of test coverage.

I frame this differently. There are different types of tests: unit tests, integration, mutation, fuzzing, etc. They all have coverage. The unit tests in all of unified have 100% test coverage. Adding mutation tests don’t increase that number. It’s a different metric. It adds more tests.

> If a contributor came through and said "line coverage is confusing, so I didn't add it" you'd close the PR without further discussion.

Yes. I don’t want to do that for mutation tests.
Adding more tools doesn’t help this point. Not adding more tools helps this. Having some tools already doesn’t change this.

> It improves readability [... and the rest]

I disagree on `for .. of` loops increasing readability over `while`.

> rather than pent up frustration with mdx

It is an example. It is a different example from what you are responding to though. In 2020, the MDX repo was unusable.

> too much tooling: this is highly subjective and I respectfully disagree

I was explaining problems with MDX 2 years ago. But to expand on Stryker:
- Mutation testing is not common or standard practise compared to say unit tests or coverage.
- OG `stryker` and styker/core are used in very few repos after 5 years.
- It doesn’t really help in this PR. The code is fine. A while loop is changed.

> too slow: it takes a minute and thirty seconds total for all tests including mutation tests

This was about MDX. For Stryker: stryker/core has a 30mb install size. It does run the tests 100(s) of times so that costs some time.
This was mostly about adding several, arbitrary tools though.

> constantly broken: this is a type of test coverage, if the code isn't covered by tests, the build should be broken 🤔

This was in reference to MDX. For Stryker: i’m worried that it’ll start emitting an unreadable message for every (other) line that’s added.
This was mostly about adding several, arbitrary tools though, because tools break.